### PR TITLE
Propose data generation schema [WIP]

### PR DESCRIPTION
**Pull request by 1602**
This PR is not for merging, yet, just discussion of proposal.

Reason for that: generalise data generation by making it configurable via protocol.

### CODE REVIEW

**Comment by qubyte** [comment]
I don't object to the approach, but I'm skeptical that it can be used in certain circumstances. For example, how does this handle addresses or vehicle registration numbers? We pick these particular things from pre-generated lists right now, but that's a detail the generation schema probably shouldn't be aware of.

File: schema/motor-insurance.json [hunk 1]
```diff
@@ -27,7 +27,26 @@
...
         },
         "vehicle": {
-            "typeRef": "#/domains/MotorInsurance/types/Vehicle"
+            "typeRef": "#/domains/MotorInsurance/types/Vehicle",
+            "dataGenerationInstruction": {
+                "method": "DataGenerationApi",
+                "generatorRequestSchema": {
+                    "type": "object",
+                    "properties": {
+                        "vehicle": {
```
Comment (1602) [inline schema/motor-insurance.json]: @qubyte this project is built based on this spec: https://github.com/universalbasket/data-generation/blob/master/SPEC.md#post-domainsmotorinsurancetypesvehicle

I think of this as a boundary between involved parties, so it doesn't really belongs to neither of data generation service nor its client (job input bundler). Both client and server need to know about this schema, server in order to validate it before saving, client in order to validate it before sending to service api.

**Comment by inca** [comment]
I've no major objections here — just a general concern that this becomes a God-class-like component.

What that means is: obviously the Domain entity is a point of interest for various actors. I can see it is tempting to put various "per-domain" entities and or configuration directly to protocol — as opposed to have pattern matching based on protocol in various places.

However, this might not be the best course of action here, because the entity itself (the Domain) becomes increasingly complex (conceptually, i.e. with all new terms like "instruction", "method", "data generation", "job", "bundling", etc, etc) and a centerpiece data communication point between components. In other words, everything we put into Domain makes its interface heavier, and this affects all existing consumers, not just few of them which need some shared knowledge.

I believe the correct course of action would be to allocate a different entity, either in the `protocol` repo, or in service-api or elsewhere (separate topic), which itself depends on the protocol, links the domain and provides more specific metadata for specific consumers.

This is the approach we already taken with Autopilot, for example, who does not store domain-specific pipes, recipes and inspections in protocol (because obviously they don't make much sense to other protocol consumers) and instead has a [dedicated repo](https://github.com/universalbasket/a-team) which follows some simple conventions to pick up the domain name. This way it is still possible for other consumers to "not know" and to "not care" about Autopilot-specific terminology like pipes and inspections.

I think honestly think it's best to keep the public interface of this "core" dependency to minimum, so I believe there should be some reasonable way of achieving that. I also think some diagramming would definitely help here, just to understand how many new terms we're bringing and what parties are actually "interested" in them.

Happy to discuss this further, but won't stay on your way in case you want to put this forward (just ack on the concern and 🚀)

**Comment by 1602** [comment]
I will reopen another PR following proposed course of actions. Thanks everyone!

**Pull request closed without merging**